community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/ExplicitIndexCursorWritesTestBase.java

```diff
@@ -0,0 +1,159 @@
+import static org.junit.Assert.assertTrue;
+
+public abstract class ExplicitIndexCursorWritesTestBase<G extends KernelAPIWriteTestSupport>
+        extends KernelAPIWriteTestBase<G>
+{
+
+    private static final String INDEX_NAME = "foo";
+    private static final String KEY = "bar";
+    private static final String VALUE = "this is it";
+
+    @Test
+    public void shouldRemoveNodeFromExplicitIndex() throws Exception
+    {
+        // Given
+        long nodeId = addNodeToExplicitIndex();
+
+        // When
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            ExplicitIndexWrite indexWrite = tx.indexWrite();
+            indexWrite.nodeRemoveFromExplicitIndex( INDEX_NAME, nodeId );
+            tx.success();
+        }
+
+        // Then
+        try ( org.neo4j.graphdb.Transaction ctx = graphDb.beginTx() )
+        {
+            IndexHits<Node> hits = graphDb.index().forNodes( INDEX_NAME ).get( KEY, "this is it" );
+            assertFalse( hits.hasNext() );
+            hits.close();
+            ctx.success();
+        }
+    }
+
+    @Test
+    public void shouldHandleRemoveNodeFromExplicitIndexTwice() throws Exception
+    {
+        // Given
+        long nodeId = addNodeToExplicitIndex();
+
+        // When
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            ExplicitIndexWrite indexWrite = tx.indexWrite();
+            indexWrite.nodeRemoveFromExplicitIndex( INDEX_NAME, nodeId );
+            tx.success();
+        }
+
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            ExplicitIndexWrite indexWrite = tx.indexWrite();
+            indexWrite.nodeRemoveFromExplicitIndex( INDEX_NAME, nodeId );
+            tx.success();
+        }
+
+        // Then
+        try ( org.neo4j.graphdb.Transaction ctx = graphDb.beginTx() )
+        {
+            IndexHits<Node> hits = graphDb.index().forNodes( INDEX_NAME ).get( KEY, "this is it" );
+            assertFalse( hits.hasNext() );
+            hits.close();
+            ctx.success();
+        }
+    }
+
+    @Test
+    public void shouldRemoveNonExistingNodeFromExplicitIndex() throws Exception
+    {
+        // Given
+        long nodeId = addNodeToExplicitIndex();
+
+        // When
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            ExplicitIndexWrite indexWrite = tx.indexWrite();
+            indexWrite.nodeRemoveFromExplicitIndex( INDEX_NAME, nodeId + 1 );
+            tx.success();
+        }
+
+        // Then
+        try ( org.neo4j.graphdb.Transaction ctx = graphDb.beginTx() )
+        {
+            IndexHits<Node> hits = graphDb.index().forNodes( INDEX_NAME ).get( KEY, VALUE );
+            assertThat( hits.next().getId(), equalTo( nodeId ) );
+            assertFalse( hits.hasNext() );
+            hits.close();
+            ctx.success();
+        }
+    }
+
+    @Test
+    public void shouldCreateExplicitIndex() throws Exception
```

Comment:
Can we also have test for creating same explicit index twice?

Reply:
Done